README.md
**Miyako for Web (M4W)**

version 0.0.1

2013.07.02 Cyross Makoto

# はじめに

本ソフトは、HTML5ベースのゲーム開発を支援するjQueryプラグインです。

# 必要環境

 1. Webサーバ
  * ApacheでもRuby on RailsでもSinatraでもnode.js...etc
  * もしくは、Phonegapなどの、HTML5/Javascriptを使用するハイブリッドアプリ開発ツール
  * 推奨:Ruby on Rails 3.1以降が動く環境
    できれば3.2以降
 2. jQuery
  * 1.9.0/2.0.0以降
  * jQuery UIなどはお好みで(名前が被らないように!)
 3. HTML、画像、CSSなどのファイル(おこのみで)

# ファイル構成

ダウンロードしたアーカイブファイルを展開すると、以下のファイル・ディレクトリが作成されます。

 * doc ドキュメントディレクトリ
 * sample サンプルディレクトリ
 * README.md 本ファイル
 * NEWS.txt 更新履歴
 * jquery.m4w.js M4W本体
 * jquery.m4w.sound.js M4Wを使用し、音声を扱うための拡張プラグイン
 * jquery.m4w.video.js M4Wを使用し、ビデオを再生するための拡張プラグイン
 * jquery.m4w.input.js M4Wを使用し、マウスカーソルの位置などの入力情報を扱うための拡張プラグイン
 * jquery.m4w.sprite_ex.js M4Wを使用し、強力なスプライトを扱うための拡張プラグイン

# 最低限の前準備(よくある方法)：
 1. HTMLファイルを作成する
 2. HTML内で、どこにm4wの情報を表示するか決める
 3. HTML上でjQuery-X.Y.Z.jsのロードを追加する(jQuery-X.Y.Z.min.jsでも可)
 4. HTML上でjQuery.m4w.jsのロードを追加する
 5. jQueryオブジェクトのm4wメソッドを呼び出す
   * 例1・bodyブロック内
       $("body").m4w();
   * 例2・divブロック内(id=area1)
       $("div#area1").m4w();

実際の例は以下

[index.html]

    <DOCTYPE HTML>
    <html>
      <head>
        <meta charset=utf-8>
        <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
        <title>Miyako 4 Web sample 0001</title>
        <link rel="stylesheet" href="main.css" type="text/css" />
        <script type="text/javascript" src="jquery-2.0.0.js"></script>
        <script type="text/javascript" src="jquery.m4w.js"></script>
        <script type="text/javascript" src="main.js"></script>
      </head>
      <body>
        <h2>10秒後に止まります</h2>
        <p>ボタンをクリックすると、少しずつスプライトが下に移動します</p>
        <div id="message1"></div>
        <div id="message2"></div>
        <div id="game"></div>
        <div id="button1"></div>
      </body>
    </html>

[main.css]

    div#game {
      background: #008000;
      position: absolute;
      top: 192px;
      left: 0;
      width: 640;
      height: 480;
    }

    div#button1 {
      background-image: url('img/clickme.gif');
      width: 128;
      height: 48;
      position: absolute;
      top: 128px;
      left: 0;
    }

[main.js]

    $(document).ready(function(){
      // メインロジック
      var main_logic = function(){
        window.m4w.screen.sprites[1].move({dx:1,dy:0});
      };

      // アセットロード終了時の処理
      // ・メインロジックの登録
      // ・ボタン押下次の処理の追加
      var on_ready = function(assets){
        window.m4w.screen.sprites.push(assets.sprite.s01);
        window.m4w.main_logic = main_logic;
        $("#button1").on("click", function(){
          window.m4w.screen.sprites[1].move({dx:0,dy:4});
        });
      }

      // M4Wの初期化
      var body = $("#game");
      body.m4w();

      // タイマー・カウント表示をスプライトとして実装
      // (本来は、メインロジックに入れるべきだが、スプライトの応用例として実装)
      window.count = 0;
      body.m4w.screen.sprites.push({
        render: function(ctx){
          window.count++;
          $("#message2").html("Time: "+(new Date()).getTime() + "/ Count: " + window.count);
        }
      });

      // スタート時の時間を表示
      $("#message1").html("Time: "+(new Date()).getTime());
      // アセットのロード
      AssetsLoader.load({
        assets: [{type: "sprite", id: "s01", src:"img/sampleimage.gif"}],
        success: on_ready
      });

      // メインループの開始
      window.m4w.main_loop();

      // 10秒後にメインループを終了するように設定
      window.setTimeout(window.m4w.stop_main_loop, 10000);
    });

# リファレンスマニュアル

展開した時に同梱される doc ディレクトリをご参照ください
`docディレクトリのドキュメントは、「JsDoc Toolkit v2.4.0」を使用しています。`

# ラインセンス

本プログラムの使用・再配布に関してMIT Lisenceを適応します。

> The MIT License (MIT)
> Copyright (c) 2012-2013 Cyross Makoto
>

NEWS.txt
更新履歴

2012/04/20
・初回アップデート

2012/04/29
・一部バグ修正
・formに関数追加
・サンプルゲーム「ライフゲーム」追加
・サンプルゲーム「ライフゲームv2」追加

2012/04/30
・一部バグ修正
・formに関数追加
・サンプルゲーム「ライフゲームv3」追加

2012/05/01
・form系関数の一部リファクタリング
・form系関数のオプションにclasses(HTMLブロックのclass属性)を追加
・ブラウザがサポートされているかの判別関数を追加(window.m4w.is_support)
・スマートフォンからの接続かどうかの判別関数を追加(window.m4w.is_smart_phone)

2013/05/18-19
・大幅な仕様変更
　・jquery.m4w.form.jsの削除
　・Layer,Background,M4WDOMクラスの削除
　・ThredsクラスをThreadManagerクラスに変更
　・RendererクラスをScreenRender,SpriteRenderクラスに分割
　・$(...).m4w系関数の対象を「全ブロック」から、「最初のブロック」に集約
　・スレッド管理の簡略化(IDによる管理から配列による管理に)
　・スプライト管理の簡略化(IDによる管理から配列による管理に)
　・prerender、postrenderの削除(全てspritesで管理)
　　・配列の要素のオブジェクトにrender(ctx)メソッドを追加することで対処
　・処理の簡略化
　・Sprite、Drawerからv属性を削除
　　・表示・非表示を選択するときは、Screen.spritesのインデックスを負数にするか、文字列にして下さい
　・既存のSpriteクラスをSpriteExクラスに変更
　　・jquery.m4w.sprite_ex.jsに移動
　　・シンプルなスプライトをSpriteクラスとする
　　・SpriteExRendererクラスを追加
　・Drawerクラスの簡略化
　　・登録できる処理関数はひとつのみ(パラメータ名も"drawers"から"func"に変更)
　・Drawerクラスのdrawメソッドをrenderメソッドに変更
　　・Spriteに属する形に変更
　・Sooundクラスをjquery.m4w.sound.jsに分離
　・Thread,ThreadManagerクラスをjquery.m4w.thread.jsに分離
　　・ThreadManagerの管理をm4wでの管理から実装者による管理に変更
　・描画処理の変更
　　・描画開始関数をScreenクラスに移動
　　・描画停止関数を追加
　・画面描画のintervalをwindow.m4wに移動
　・ひとつのWindowに対して、ひとつのScreenに制限
　　2つ以上作ろうとするとエラー発生
　・特定のクラスでなくても、render(ctx)を実装しているオブジェクトならなんでもScreenRendererの対象になった
　・Thread.check、Thread.waitingメソッドの廃止
　　・checkはThread.updateメソッドの返却値で代用する
　・Thread.pause、Thread.resumeメソッドの廃止
　・githubリポジトリ内容の置き換え

2013/05/20
・サンプル(0001)の追加
・ドキュメントの修正

2013/05/21
・Drawerクラスを、jquery.m4w.drawer.jsに移動
・window.m4w.create_canvasメソッドを追加
・Screenコンストラクタにzパラメータを復活
・サンプル0001がIE10で正しく表示されない問題を修正
・M4W.varsをwindow.m4w.varsに移動
・window.m4w.input_varsを追加
・サンプル0002を追加

2013/05/22
・サンプル0003を追加

2013/05/23
・window.m4w.main_loop、window.m4w._game_loop、window.m4w.stop_main_loopの
　呼び出しタイミングによっては、screenなどが取得できない問題を修正
・Drawerの修正
・サンプル0004を追加
・Screen.resizeメソッドのバグを修正

2013/05/25
・jquery.m4w.thread.jsを削除

2013/05/27

jquery.m4w.js
/**
 * @fileOverview Miyako4Web(Miyako for Web a.k.a M4W/m4w)本体<br>
 * Thanks Kudox.jp(http://kudox.jp/html-css/html5-canvas-animation)<br>
 * Thanks Asial blog(http://blog.asial.co.jp/883)
 *
 * @name Miyako for Web(M4W)
 * @author Cyross Makoto (サイロス誠)
 * @version 0.0.1
 * @revision 1
 * @require jQuery 1.9.0/2.0.0(or later)
 * @license MIT License (MITライセンス)
 */

(function($){

  /**
   * @constructor
   * @class アセット(画像、スプライト、音声など)をブラウザに読み込む
   */
  AssetsLoader = function(){
  };

  /** @ignore */
  AssetsLoader.loaders = {
    /** @ignore */
    image: function(options){ return Sprite.load_image(options); },
    /** @ignore */
    sprite: function(options){ return Sprite.load(options); }
  };

  /**
   * 画像などのアセットを読み込み、指定した関数をコールバックする<br>
   * すべてのロードに成功したときはsuccess関数、一つでも失敗したときはfailed関数を呼ぶ
   * success関数の引数は以下のキーを持つ辞書オブジェクト<br>
   * <table>
   * <tr><th>Key</th><th>Value</th></tr>
   * <tr><td>image</td><td>Imageオブジェクトの辞書オブジェクト</tr>
   * <tr><td>sprite</td><td>Spriteオブジェクトの辞書オブジェクト</tr>
   * </table>
   * failed関数の引数はdefer.when.failedメソッドの引数の配列が渡される<br>
   * @example 実際のImage/Sprite/Soundを取得するときは以下の方法でアクセス可能<br>
   * 例1)<br>
   * function success(assets){<br>
   *    var image = assets.image.(id);<br>
   *    var sprite = assets.sprite.(id);<br>
   * }<br>
   * 例2)<br>
   * function success(assets){<br>
   *    var image = assets["image"]["id"];<br>
   *    var sprite = assets["sprite"]["id"];<br>
   * }
   * @param options.assets アセット設定オブジェクトの配列
   * @param options.assets[].type アセットの形式<br>画像は"image"、スプライトは"sprite"と指定する
   * @param options.assets[].その他 Image,Spriteクラスコンストラクタの引数と同じものを指定
   * @param [options.success] ロードが全て完了したときに呼び出されるコールバック関数<br>引数はアセット管理オブジェクト<br>内容は後述
   * @param [options.failed] ロードに失敗した時に呼び出されるコールバック関数<br>引数は失敗した時に渡される引数リスト
   */
  AssetsLoader.load = function(options){
    var o = $.extend({
      success: function(assets){ return assets; },
      failed: function(args){ alert("error occured!"); return args; }
    }, options);

    var assets = {}

    for(var i=0, len=o.assets.length; i<len; i++){
      var asset = o.assets[i]
      assets[asset.id] = asset;
    }

    var exec_assets = [];
    for(var asset_id in assets){
      var asset = assets[asset_id];
      exec_assets.push((AssetsLoader.loaders[asset.type])(asset));
    }
    $.when.apply(this, exec_assets).then(function(){
      var loaded_assets = {
      };
      for(var i=0, alen=arguments.length; i<alen; i++){
        var asset = arguments[i];
        if (!(asset.type in loaded_assets)){ loaded_assets[asset.type] = {}; }
        loaded_assets[asset.type][asset.id] = asset.value;
      }
      o.success(loaded_assets);
    }).fail(function(){
      o.failed(arguments);
    });
  };

  /**
   * @constructor
   * @class 画面用レンダリングメソッドを定義
   */
  ScreenRenderer = function(){
    this.name= "Miyako4Web Renderer";
  };

  /**
   * 画面への描画を行う<br>
   * 次の順番で描画する
   * <ul>
   * <li>画面の内容を消去する</li>
   * <li>各スプライトを描画する</li>
   */
  ScreenRenderer.default_render = function(){
    var ctx = this.context;
    var ia = this.sprites;
    ctx.clearRect(0, 0, this.width, this.height);
    for(var i=0, ialen=ia.length; i<ialen; i++){
      var s = ia[i];
      if(s == null){ continue; }
      if(!('render' in s)){ continue; }
      s.render(ctx);
    }
    return this;
  };

  /**
   * 画面の消去を行わない画面への描画を行う<br>
   * 次の順番で描画する
   * <ul>
   * <li>各スプライトを描画する</li>
   */
  ScreenRenderer.no_clear_render = function(){
    var ctx = this.context;
    var ia = this.sprites;
    for(var i=0, ialen=ia.length; i<ialen; i++){
      var s = ia[i];
      if(s == null){ continue; }
      if(!('render' in s)){ continue; }
      s.render(ctx);
    }
    return this;
  };

  /**
   * 画面の描画を行わない
   */
  ScreenRenderer.no_render = function(){
    return this;
  };

  /**
   * @constructor
   * @class スプライト用レンダリングメソッドを定義
   */
  SpriteRenderer = function(){
    this.name= "Miyako4Web Renderer";
  };

  /**
   * 画面への描画を行う<br>
   * 隠蔽中の時は描画しない<br>
   * 次の順番で描画する
   * <ul>
   * <li>画像を描画する(image)</li>
   * </ul>
   */
  SpriteRenderer.default_render = function(context){
    context.globalAlpha = this.a;
    context.drawImage(this.image, this.x, this.y);
  };

  /**
   * @constructor
   * @class 画面(ブロック要素領域)・画面上で発生するイベントを管理<br>スプライトの描画制限は、sprites配列の要素のインデックスをマイナスにするか、数値以外の文字列にする
   * @param options.id 画面のID
   * @param [options.width] 画面の横幅<br>ピクセル単位<br>省略時は640
   * @param [options.height] 画面の高さ<br>ピクセル単位<br>省略時は480
   * @param [options.position] bodyからの位置関係(CSSのpositionと同じ)<br>省略時は"relative"
   * @param [options.left] ブラウザ(もしくは親ブロックからの右方向の位置<br>単位はピクセル<br>省略時は0(ブロックの左端)
   * @param [options.top] ブラウザ(もしくは親ブロックからの上方向の位置<br>単位はピクセル<br>省略時は0(ブロックの上端)
   * @param [options.body] 画面レイヤーを埋め込むブロック<br>jQueryオブジェクトを指定<br>省略時はbodyブロック($("body"))
   * @param [options.css] CSS値をオブジェクトで指定<br>省略時は{}
   * @param [options.attr] attr値をオブジェクトで指定<br>jQueryオブジェクトを指定<br>省略時は{}
   * @param [options.events] 登録したいイベントの関数群<br>"イベント名": イベントハンドラ本体で登録する<br>省略時は{}
   * @param [options.content_width] 描画領域の横幅<br>単位はピクセル<br>指定したときは、widthで指定した範囲まで拡大して表示される<br>省略時はwidthと同じ値
   * @param [options.content_height] 描画領域の高さ<br>単位はピクセル<br>指定したときは、heightで指定したはんいまで拡大して表示される<br>省略時はheightと同じ値
   * @param [options.renderer] 描画する際の処理関数<br>省略時はRenderer.default_render
   * @param [options.z] ブラウザ上の前後位置を指定する<br>省略時は0
   * @param [options.get_context] レイヤーのコンテキストを作成するかどうかのフラグ<br>WebGLを使う時など、コンテキストを作成する必要がないときはfalseを渡す<br>falseを渡したときは、画面のレンダラが自動的にScreenRenderer.no_renderメソッドが指定される<br>(trueの時はScreenRenderer.default_render)<br>省略時はtrue
   */
  Screen = function(options){
    var o = $.extend({
      width: 640,
      height: 480,
      position: "absolute",
      left: 0,
      top: 0,
      renderer: ScreenRenderer.default_render,
      body: $("body"),
      events: {},
      z: 0,
      get_context: true
    }, options);

    var cw = o.width;
    var ch = o.height;
    if("content_width" in o){ cw = o.content_width; }
    if("content_height" in o){ ch = o.content_height; }

    var $layer = window.m4w.create_canvas(o);

    $layer.appendTo(o.body);

    this.body = o.body;
    /**
     * @property 画面幅
     */
    this.width = o.width;
    /**
     * @property 画面高さ
     */
    this.height = o.height;
    /**
     * @property 描画領域幅
     */
    this.content_width = cw;
    /**
     * @property 描画領域高さ
     */
    this.content_height = ch;
    /**
     * @property 画面レイヤ
     */
    this.layer = $layer;
    /**
     * @property 画面ID
     */
    this.id = o.id;
    /**
     * @property 画面のデバイスコンテキスト(必要なとき)
     */
    if(o.get_context){
      this.context = this.layer[0].getContext("2d");
    }
    else{
      this.context = null;
      o.renderer = ScreenRenderer.no_render;
    }
    /**
     * @property スプライト配列
     */
    this.sprites = [];
    /**
     * @property 画面レンダラ
     */
    this.render = o.renderer.bind(this);

    for(name in o.css){ $layer.css(name, o.css[name]); }
    for(name in o.attr){ $layer.attr(name, o.attr[name]); }
    for(name in o.events){
      $screen.unbind(name);
      $screen.bind(name, o.events[name]);
    }
  };

  /** 画面の内容を消去する */
  Screen.prototype.clear = function(){
    this.context.clearRect(0, 0, this.width(), this.height());
  };

  /** 画面の横幅・高さを変更する
   * @param {int} width 横幅
   * @param {int} height 高さ
   */
  Screen.prototype.resize = function(width, height){
    var w = parseInt(width);
    var h = parseInt(height);
    this.width = w;
    this.height = h;
    this.layer.css("width", w);
    this.layer.css("height", h);
    this.layer[0].width = (typeof content_width !== 'undefined' ? parseInt(content_width) : w);
    this.layer[0].height = (typeof content_height !== 'undefined' ? parseInt(content_height) : h);
  };

  /** 画面の位置を変更する
   * @param {int} left 右方向の位置
   * @param {int} top 下方向の位置
   */
  Screen.prototype.move = function(left, top){
    this.body.css("left", left);
    this.body.css("top", top);
  };

  /**
   * @constructor
   * @class スプライトを管理・操作
   * @param options.id スプライトに一意に一位につけられるID<br>省略時はvalue属性(Imageオブジェクト)のidプロパティ
   * @param options.value スプライトの元にするImageクラス(Javascript標準クラス)のオブジェクト
   * @param [options.x] ブロックの左端から右方向の位置(左端を0とする)<br>省略時は0
   * @param [options.y] ブロックの左端から右方向の位置(左端を0とする)<br>省略時は0
   * @param [options.a] 描画時の透明度<br>0.0≦a≦1.0の間<br>0.0で完全透明、1.0で完全不透明<br>省略時は1.0
   */
  Sprite = function(options){
    var o = $.extend({
      id: options.value.id,
      x:0, y:0, a:1.0
    }, options);
    this.id = o.id;
    this.image = o.value;
    /** @property ブロックの左端から右方向の位置(左端を0とする) */
    this.x = o.x;
    /** @property ブロックの左端から右方向の位置(左端を0とする) */
    this.y = o.y;
    /** @property 描画時の透明度<br>0≦a≦1の間<br>0で完全透明、1で完全不透明 */
    this.a = o.a;
  };

  /**
   * 画像ファイルを読み込む<br>imgタグを作り、再生ができるときに指定した関数を呼び出す
   * @param options.id 画像に一意につけられるID
   * @param options.src 対象ファイルのURLを配列で指定(マルチブラウザ対応)
   * @return 画像を読み込んでいるDeferredオブジェクト<br>コールバック関数の引数は、以下の内容の示すオブジェクト
   * <ul>
   *  <li>type: "image"</li>
   *  <li>id: options.id</li>
   *  <li>value: 生成したImageオブジェクト</li>
   *  <li>options: 生成時のオプション</li>
   * </ul>
   */
  Sprite.load_image = function(options){
    var defer = $.Deferred();
    var img_id = options.id;
    var img = new Image();
    /** @ignore */
    img.src = options.src+ "?" + new Date().getTime();
    /** @ignore */
    img.onload = function(){ defer.resolve({type: "image", id: img_id, value: img, options: options}); };
    return defer.promise();
  };

  /**
   * 画像をロードし、読み込み終了後Spriteオブジェクトを作成する
   * @param options.id スプライトに一意に一位につけられるID<br>内部で生成するImageオブジェクトも同じIDになる
   * @param options.src 対象ファイルのURLを配列で指定(マルチブラウザ対応)
   * @return スプライトを生成しているDeferredオブジェクト<br>コールバック関数の引数は、以下のパラメータは以下の内容のオブジェクト
   * <ul>
   *  <li>type: "sprite"</li>
   *  <li>id: options.id</li>
   *  <li>value: 生成したSpriteオブジェクト</li>
   *  <li>options: 生成時のオプション</li>
   * </ul>
   */
  Sprite.load = function(options){
    var defer = $.Deferred();
    Sprite.load_image(options).then(function(obj){
      defer.resolve({type: "sprite", id: obj.id, value: new Sprite($.extend(options, obj)), options: options});
    });
    return defer.promise();
  };

  /**
   * スプライトを描画する<br>省略時はRenderer.default_render関数が指定される
   * @param {Object} context 対象のレイヤー(canvasタグ)が持つコンテキスト
   */
  Sprite.prototype.render = SpriteRenderer.default_render;

  /**
   * スプライトを移動させる<br>座標は、ブロックの左上が(0,0)で、値がプラスだと右・下方向、マイナスだと左・上方向となる
   * @param [options.x] 左端を0とした時の右方向の位置<br>省略時は現在位置
   * @param [options.y] 上端を0とした時の下方向の位置<br>省略時は現在位置
   * @param [options.dx] 右方向をプラスとした時の移動量<br>省略時は0
   * @param [options.dy] 下方向をプラスとした時の移動量<br>省略時は0
   */
  Sprite.prototype.move = function(options){
    var o = $.extend({x: this.x, y: this.y, dx: 0, dy: 0}, options);
    this.x = o.x + o.dx;
    this.y = o.y + o.dy;
  };

  /**
   * 描画開始関数を取得(内部で使用)<br>
   * Thanks Kudox.jp(http://kudox.jp/html-css/html5-canvas-animation)<br>
   */
  window.requestAnimFrame = (function(){
    return window.requestAnimationFrame       ||
           window.webkitRequestAnimationFrame ||
           window.mozRequestAnimationFrame    ||
           window.oRequestAnimationFrame      ||
           window.msRequestAnimationFrame     ||
           function(/* function */ callback){ return window.setTimeout(callback, 1000/window.m4w.interval); };
  }());

  /**
   * 描画停止関数を取得(内部で使用)<br>
   * Thanks Kudox.jp(http://kudox.jp/html-css/html5-canvas-animation)<br>
   */
  window.cancelAnimFrame = (function() {
    return window.cancelAnimationFrame ||
           window.cancelRequestAnimationFrame ||
           window.webkitCancelAnimationFrame ||
           window.webkitCancelRequestAnimationFrame ||
           window.mozCancelAnimationFrame ||
           window.mozCancelRequestAnimationFrame ||
           window.msCancelAnimationFrame ||
           window.msCancelRequestAnimationFrame ||
           window.oCancelAnimationFrame ||
           window.oCancelRequestAnimationFrame ||
           function(id) { window.clearTimeout(id); };
  }());

  /**
   * @param options.body 処理対象のjQueryオブジェクト
   * @constructor
   * @class $.fn.m4w呼び出し後の各種処理をまとめる
  */
  M4W = function(options){
    this.body = options.body;
    // screen_optionsに"body"がないときは、this.bodyを追加
    if(!("screen_options" in options)){
      options["screen_options"] = {body: this.body};
    }
    else if(!("body" in options.screen_options)){
      options.screen_options["body"] = this.body;
    }
    /**
     * @property スクリーン共有オブジェクト<br>Screenクラス
     */
    this.screen = new Screen(options.screen_options);
    if(!(typeof window.m4w.screen === 'undefined')){
      throw new Error("Screen object is already registered in window.m4w!");
    }
    window.m4w.screen = this.screen;
  };

  /**
   * 対象のDOMオブジェクトにアセットを追加する
   * @param {Object} options AssetsLoader.loadメソッドの引数と同じ
   */
  M4W.prototype.add_assets = function(options){
    AssetsLoader.load(options);
    return this.body;
  };

  /**
   * windowオブジェクト
   * @name window
  */

  /**
   * @namespace それぞれのクラスのアクセス用名前空間<br>以下のクラスが利用可能
   * <ul>
   * <li>Screen</li>
   * <li>AssetsLoader</li>
   * <li>Sprite</li>
   * <li>SpriteRenderer</li>
   * <li>ScreenRenderer</li>
   * </ul>
   * また、以下の関数が利用可能
   * <ul>
   * <li>user_agent</li>
   * </ul>
   * また、以下のプロパティが利用可能
   * <ul>
   * <li>interval(画面の描画間隔、初期値は60(60fps相当)、ただし、間隔取得にsetTimeoutを使用するときのみ)</li>
   * <li>vars(データ共有オブジェクト　初期は空のオブジェクト)</li>
   * <li>input_vars(入力デバイス関連情報共有オブジェクト<br>初期は空のオブジェクト)</li>
   * </ul>
  */
  window.m4w = {
    Screen: Screen,
    AssetsLoader: AssetsLoader,
    Sprite: Sprite,
    SpriteRenderer: SpriteRenderer,
    ScreenRenderer: ScreenRenderer,
    /**
     * @property ユーザーエージェントを返す関数<br>
     * 返されるオブジェクトは、user_agent, versionの2つのプロパティを持つ<br>
     */
    user_agent: function(){
        var result = {};
        result.user_agent = window.navigator.userAgent.toLowerCase();
        result.version = window.navigator.appVersion.toLowerCase();
        return result;
      },
    /**
     * @property 画面更新間隔(初期値:60(fps))
     */
    interval: 60,
    /**
     * @property 画面更新ID(内部で使用)
     */
    rendering_id: null,
    /**
     * @property 画面更新終了フラグ(内部で使用)
     */
    is_stop: false,
    /**
     * @property ゲームロジック
     */
    main_logic: function(){ },
    /**
     * @property データ共有オブジェクト<br>初期は空のオブジェクト
     */
    vars: {},
    /**
     * @property 入力デバイス関連情報共有オブジェクト<br>初期は空のオブジェクト
     */
    input_vars: {},
    /**
     * @propert メインループ実行中を示すフラグ。trueの時にメインループに入ると処理を飛ばす
     */
    is_main_loop: false
  };

  /**
   * ゲームループ内部ロジック
   */
  window.m4w._game_loop = function(){
    var _m4w = window.m4w;
    if(_m4w.is_main_loop){ return; }
    _m4w.is_main_loop = true;
    // メインロジック実行
    _m4w.main_logic();
    // 外部から停止リクエストが来たときは終了
    if(_m4w.is_stop){
      window.cancelAnimFrame(_m4w.rendering_id);
      _m4w.is_stop = false;
      _m4w.is_main_loop = false;
      return;
    }
    // 画面描画
    _m4w.screen.render();
    // 次のループ実行を設定
    _m4w.rendering_id = window.requestAnimFrame(_m4w._game_loop.bind(_m4w));
    _m4w.is_main_loop = false;
  };

  /**
   * ゲームループ開始
   */
  window.m4w.main_loop = function(){
    var _m4w = window.m4w;
    _m4w.rendering_id = window.requestAnimFrame(_m4w._game_loop.bind(_m4w));
  };

  /**
   * ゲームループを停止する
   */
  window.m4w.stop_main_loop = function(){
    window.m4w.is_stop = true;
  };

  /**
   * Canvasタグを作成する
   * @param options.id 画面のID
   * @param [options.width] 画面の横幅<br>ピクセル単位<br>省略時は640
   * @param [options.height] 画面の高さ<br>ピクセル単位<br>省略時は480
   * @param [options.position] bodyからの位置関係(CSSのpositionと同じ)<br>省略時は"relative"
   * @param [options.left] ブラウザ(もしくは親ブロックからの右方向の位置<br>単位はピクセル<br>省略時は0(ブロックの左端)
   * @param [options.top] ブラウザ(もしくは親ブロックからの上方向の位置<br>単位はピクセル<br>省略時は0(ブロックの上端)
   * @param [options.content_width] 描画領域の横幅<br>単位はピクセル<br>指定したときは、widthで指定した範囲まで拡大して表示される<br>省略時はwidthと同じ値
   * @param [options.content_height] 描画領域の高さ<br>単位はピクセル<br>指定したときは、heightで指定したはんいまで拡大して表示される<br>省略時はheightと同じ値
   * @param [options.z] ブラウザ上の前後位置を指定する<br>省略時は0
   */
  window.m4w.create_canvas = function(options){
    var o = $.extend({
      width: 640,
      height: 480,
      position: "absolute",
      left: 0,
      top: 0,
      z: 0
    }, options);
    var $canvas = $('<canvas />', {width: o.width, height: o.height});
    var cw = o.width;
    var ch = o.height;
    if("content_width" in o){ cw = o.content_width; }
    if("content_height" in o){ ch = o.content_height; }
    $canvas.attr("id", o.id);
    $canvas.css("position", o.position);
    $canvas.css("left", o.left);
    $canvas.css("top", o.top);
    $canvas.css("z-index", o.z);
    $canvas[0].width = cw;
    $canvas[0].height = ch;
    return $canvas;
  };

  /**
   * 引数のUser Agent (複数指定可)から、タッチデバイスかどうか判別する<br>
   * @param agent 検索したいagentの配列(全て小文字)
   * @return 指定のagentのどれかに合致した時はtrue、それ以外はfalse
   */
  window.m4w._is_device = function(agents){
    var user_agent = window.navigator.userAgent.toLowerCase();

    for(var i=0; i<agents.length; i++){
      if(user_agent.indexOf(agents[i]) > -1){ return true; }
    }
    // その他
    return false;
  };

  /**
   * User Agent から、タッチデバイスかどうか判別する(Windows 8を除く)<br>
   * 使用している端末・OSが以下の時、trueとなる
   * <ul>
   * <li>イベントにontouchstartが用意されている
   * <li>iPhone
   * <li>iPad
   * <li>iPod Touch
   * <li>Android端末
   * <li>Windows Phone
   * </ul>
   * @return タッチデバイスの時はtrue、それ以外はfalse
   */
  window.m4w.is_touch_device = function(){
    if(document.ontouchstart !== undefined){ return true; }
    return window.m4w._is_device(["iphone", "ipad", "ipod", "android", "windowsphone"]);
  };

  /**
   * User Agent から、Appleのデバイス(iPhone,iPad,iPod Touch)かどうか判別する<br>
   * @return Appleのデバイスの時はtrue、それ以外はfalse
   */
  window.m4w.is_apple_device = function(){
    return window.m4w._is_device(["iphone", "ipad", "ipod"]);
  };

  /**
   * User Agent から、iPhoneかどうか判別する<br>
   * @return iPhoneの時はtrue、それ以外はfalse
   */
  window.m4w.is_iphone = function(){
    return window.m4w._is_device(["iphone"]);
  };

  /**
   * User Agent から、iPadかどうか判別する<br>
   * @return iPadの時はtrue、それ以外はfalse
   */
  window.m4w.is_ipad = function(){
    return window.m4w._is_device(["ipad"]);
  };

  /**
   * User Agent から、iPod Touchかどうか判別する<br>
   * @return iPod Touchの時はtrue、それ以外はfalse
   */
  window.m4w.is_ipod = function(){
    return window.m4w._is_device(["ipod"]);
  };

  /**
   * User Agent から、Android端末かどうか判別する<br>
   * @return Android端末の時はtrue、それ以外はfalse
   */
  window.m4w.is_android = function(){
    return window.m4w._is_device(["android"]);
  };

  /**
   * User Agent から、使用しているOSがWindowsPhoneかどうか判別する<br>
   * @return WindowsPhoneの時はtrue、それ以外はfalse
   */
  window.m4w.is_windows_phone = function(){
    return window.m4w._is_device(["windowsphone"]);
  };

  /**
   * User Agent から、使用しているOSがWindows 8かどうか判別する<br>
   * @return Windows 8の時はtrue、それ以外はfalse
   */
  window.m4w.is_windows8 = function(){
    return window.m4w._is_device(["windows nt 6.2"]);
  };

  /**
   * jQueryオブジェクトの別名(http://jquery.com/)
   * @name $
   * @class
   */

  /**
   * jQueryオブジェクトのプラグイン関数(http://jquery.com/)
   * @name $.fn
   * @class
   */

   /**
   * 指定したブロックのm4w初期化
   * このメソッドを呼び出すと、以下の場所にオブジェクトが追加される
   * <table>
   * <tr><th>対象</th><th>名称</th><th>オブジェクト</th></tr>
   * <tr><td>jQueryオブジェクト</td><td>m4w</td><td>M4W</td></tr>
   * <tr><td>各ブロック要素のDOMオブジェクト</td><td>m4w</td><td>M4WDOM</td></tr>
   * </table>
   * @param [options.screen_options] screenオブジェクトに渡す引数<br>Screenコンストラクタの引数と同じ
   * @param [options.assets] 初期化時に作成したいアセットの配列。内容はAssetLoader.loadメソッドの引数と同じ<br>省略時はnull
   */
  $.fn.m4w = function(options){
    var o = $.extend({
      screen_options: {},
      assets: null,
      body: $(this)
    }, options);

    $.extend($(this).m4w, new M4W(o));
    if(o.assets != null){ $(this).m4w.add_assets(o.assets); }
  };
})(jQuery);
